tests/test_requests.py

```diff
@@ -221,6 +221,35 @@ def test_http_303_doesnt_change_head_to_get(self, httpbin):
         assert r.history[0].status_code == 303
         assert r.history[0].is_redirect
 
+    def test_header_and_body_removal_on_redirect(self, httpbin):
+        purged_headers = ('Content-Length', 'Content-Type', 'Transfer-Encoding')
+        ses = requests.Session()
+        req = requests.Request('POST', httpbin('post'), data={'test': 'data'})
+        prep = ses.prepare_request(req)
+
+        # Test Transfer-Encoding too
+        req_tran = requests.Request('POST', httpbin('post'), data=(b'x' for x in range(1)))
+        prep_tran = ses.prepare_request(req_tran)
+        assert 'Transfer-Encoding' in prep_tran.headers
+
+        resp = ses.send(prep)
+        resp_tran = ses.send(prep_tran)
+
+        # Mimic a redirect response
+        resp.status_code = 302
+        resp.headers['location'] = 'get'
+        resp_tran.status_code = 302
+        resp_tran.headers['location'] = 'get'
+
+        # Run request through resolve_redirects
+        next_resp = next(ses.resolve_redirects(resp, prep))
+        next_resp_tran = next(ses.resolve_redirects(resp, prep))
+        assert next_resp.request.body is None
+        assert next_resp_tran.request.body is None
+        for header in purged_headers:
+            assert header not in next_resp.request.headers
+            assert header not in next_resp_tran.request.headers
```

Comment:
Can we split this into two tests that look very similar, rather than have one really complex test?